pyfr/backends/opencl/blasext.py

```diff
@@ -43,47 +43,59 @@ def run(self, queue):
 
         return CopyKernel()
 
-    def errest(self, x, y, z, *, norm):
-        if x.traits != y.traits != z.traits:
+    def reduction(self, *rs, method, norm, dt_mat=None):
+        if any(r.traits != rs[0].traits for r in rs[1:]):
             raise ValueError('Incompatible matrix types')
 
-        nrow, ncol, ldim, dtype = x.traits[1:]
-        ncola, ncolb = x.ioshape[1:]
+        rdata = [r.data for r in rs]
+
+        nrow, ncol, ldim, dtype = rs[0].traits[1:]
+        ncola, ncolb = rs[0].ioshape[1:]
 
         # Reduction workgroup dimensions
         ls = (128, 1)
         gs = (ncolb - ncolb % -ls[0], ncola)
 
         # Empty result buffer on host with (nvars, ngroups)
-        err_host = np.empty((ncola, gs[0] // ls[0]), dtype)
+        reduced_host = np.empty((ncola, gs[0] // ls[0]), dtype)
 
         # Device memory allocation
-        err_dev = cl.Buffer(self.backend.ctx, cl.mem_flags.READ_WRITE,
-                            err_host.nbytes)
+        reduced_dev = cl.Buffer(self.backend.ctx, cl.mem_flags.READ_WRITE,
+                                reduced_host.nbytes)
+
+        tplargs = dict(norm=norm, sharesz=ls[0], method=method)
+
+        if method == 'resid':
+            tplargs['dt_type'] = 'matrix' if dt_mat else 'scalar'
 
         # Get the kernel template
-        src = self.backend.lookup.get_template('errest').render(
-            norm=norm, sharesz=ls[0]
-        )
+        src = self.backend.lookup.get_template('reduction').render(**tplargs)
+
+        # Argument types for reduction kernel
+        if method == 'errest':
+            argt = [np.int32]*3 + [np.intp]*4 + [dtype]*2
+        elif method == 'resid' and dt_mat:
+            rdata.append(dt_mat.data)
```

Comment:
Is this okay or should I get rid of this one as well?

Reply:
See which one is cleaner